components/script/dom/element.rs

```diff
@@ -856,6 +871,329 @@ impl Element {
+                        if position.x != current_scroll_x || position.y != current_scroll_y {
+                            // Step 1.3.2: Perform a scroll of the viewport to position, with root
+                            // element as the associated element
+                            window.scroll(position.x, position.y, behavior, can_gc);
+                        }
+                    },
+                }
+
+                // Step 1.4: Check container stopping condition
+                if let Some(container) = container {
+                    // If container is not null and either scrolling box is a
+                    // shadow-including inclusive ancestor of container or is a viewport
+                    // whose document is a shadow-including inclusive ancestor of
+                    // container, abort the rest of these steps.
+                    let should_stop = match scrolling_box {
+                        ScrollingBox::Element(ref element) => {
+                            // Check if the scrolling box element is a shadow-including inclusive ancestor of container
+                            element
+                                .upcast::<Node>()
+                                .is_shadow_including_inclusive_ancestor_of(
+                                    container.upcast::<Node>(),
+                                )
+                        },
+                        ScrollingBox::Viewport(ref document) => {
+                            // Check if the viewport's document is a shadow-including inclusive ancestor of container
+                            document
+                                .upcast::<Node>()
+                                .is_shadow_including_inclusive_ancestor_of(
+                                    container.upcast::<Node>(),
+                                )
+                        },
+                    };
+                    if should_stop {
+                        return;
+                    }
+                }
+            });
+    }
+
+    /// Check if an element establishes a scrolling box
+    fn establishes_scroll_box(&self, node: &Node, can_gc: CanGc) -> bool {
+        if node.is::<Document>() {
+            true // Document's viewport is always a scrolling box
+        } else if node.is::<Element>() {
+            let element = node.downcast::<Element>().unwrap();
+            if let Some(style) = element.style(can_gc) {
+                let overflow_x = style.get_box().clone_overflow_x();
+                let overflow_y = style.get_box().clone_overflow_y();
+                return overflow_x.is_scrollable() || overflow_y.is_scrollable();
+            } else {
+                false // Element without style is not a scrolling box
+            }
+        } else {
+            false // Shadow roots and other nodes are not scrolling boxes
+        }
+    }
+
+    /// <https://drafts.csswg.org/cssom-view/#element-scrolling-members>
+    fn determine_scroll_into_view_position(
```

Comment:
This function and the next one still look very similar. Can we share the common logic?